File: flutter_app/lib/services/feature_extractor.dart
import 'dart:math';
import '../models/transaction.dart';

class FeatureExtractor {
  static Map<String, dynamic> extractFeatures(List<Transaction> transactions) {
    if (transactions.isEmpty) {
      return _getDefaultFeatures();
    }

    final sortedTxns = List<Transaction>.from(transactions)
      ..sort((a, b) => a.date.compareTo(b.date));

    final incomeTransactions =
        sortedTxns.where((t) => t.type == TransactionType.credit).toList();
    final expenseTransactions =
        sortedTxns.where((t) => t.type == TransactionType.debit).toList();

    final totalIncome =
        incomeTransactions.fold(0.0, (sum, t) => sum + t.amount.abs());
    final totalExpenses =
        expenseTransactions.fold(0.0, (sum, t) => sum + t.amount.abs());

    final features = <String, double>{};

    features['transaction_count'] = sortedTxns.length.toDouble();

    final amounts = sortedTxns.map((t) => t.amount.abs()).toList();
    features['avg_transaction'] = _mean(amounts);

    final amtMean = _mean(amounts);
    final amtStd = _standardDeviation(amounts);
    features['transaction_volatility'] =
        amtMean > 0 ? amtStd / (amtMean + 1) : 0.5;

    features['expense_ratio'] =
        totalIncome > 0 ? totalExpenses / (totalIncome + 1) : 1.0;

    double maxPaymentGap = 30.0;
    if (incomeTransactions.length > 1) {
      for (int i = 1; i < incomeTransactions.length; i++) {
        final gap = incomeTransactions[i]
            .date
            .difference(incomeTransactions[i - 1].date)
            .inDays;
        if (gap > maxPaymentGap) {
          maxPaymentGap = gap.toDouble();
        }
      }
    }
    features['payment_consistency'] = maxPaymentGap;

    final overdraftCount =
        sortedTxns.where((t) => t.balance < 0).length.toDouble();
    features['overdraft_frequency'] =
        sortedTxns.isNotEmpty ? overdraftCount / sortedTxns.length : 0;

    if (incomeTransactions.length > 1) {
      final incomeAmounts = incomeTransactions.map((t) => t.amount.abs()).toList();
      final incMean = _mean(incomeAmounts);
      final incStd = _standardDeviation(incomeAmounts);
      features['income_stability'] = incMean > 0 ? incStd / (incMean + 1) : 0.5;
    } else {
      features['income_stability'] = 0.5;
    }

    final uniqueCategories = sortedTxns.map((t) => t.category).toSet().length;
    features['category_diversity'] = min(uniqueCategories / 10, 1.0);

    double accountAge = 30.0;
    if (sortedTxns.length > 1) {
      accountAge = sortedTxns.last.date
          .difference(sortedTxns.first.date)
          .inDays
          .toDouble();
      if (accountAge < 1) accountAge = 30.0;
    }
    features['account_age'] = accountAge;

    final balances = sortedTxns.map((t) => t.balance).toList();
    features['avg_balance'] = _mean(balances);

    return _mapToModelFeatures(features, totalIncome, totalExpenses);
  }

  static Map<String, dynamic> _mapToModelFeatures(
    Map<String, double> features,
    double totalIncome,
    double totalExpenses,
  ) {
    final estimatedLimit = max(totalIncome * 5, 20000.0);

    final maxGap = features['payment_consistency'] ?? 30.0;
    int paymentStatus;
    if (maxGap <= 35) {
      paymentStatus = -1;
    } else if (maxGap <= 45) {
      paymentStatus = 0;
    } else if (maxGap <= 70) {
      paymentStatus = 1;
    } else if (maxGap <= 100) {
      paymentStatus = 2;
    } else {
      paymentStatus = min((maxGap / 30).floor(), 8);
    }

    final billAmt = totalExpenses > 0
        ? totalExpenses / 3
        : (features['avg_balance'] ?? 5000) * 0.5;
    final payAmt = totalIncome > 0 ? totalIncome / 3 : billAmt * 0.3;

    return {
      'LIMIT_BAL': estimatedLimit,
      'AGE': 35.0,
      'PAY_0': paymentStatus.toDouble(),
      'PAY_2': paymentStatus.toDouble(),
      'PAY_3': paymentStatus.toDouble(),
      'BILL_AMT1': billAmt,
      'BILL_AMT2': billAmt * 0.95,
      'BILL_AMT3': billAmt * 0.9,
      'PAY_AMT1': payAmt,
      'PAY_AMT2': payAmt * 0.95,
      'PAY_AMT3': payAmt * 0.9,
      '_raw_features': features,
    };
  }

  static List<double> getFeatureVector(Map<String, dynamic> features) {
    return [
      features['LIMIT_BAL'] as double,
      features['AGE'] as double,
      features['PAY_0'] as double,
      features['PAY_2'] as double,
      features['PAY_3'] as double,
      features['BILL_AMT1'] as double,
      features['BILL_AMT2'] as double,
      features['BILL_AMT3'] as double,
      features['PAY_AMT1'] as double,
      features['PAY_AMT2'] as double,
      features['PAY_AMT3'] as double,
    ];
  }

  static Map<String, dynamic> _getDefaultFeatures() {
    return {
      'LIMIT_BAL': 50000.0,
      'AGE': 35.0,
      'PAY_0': 0.0,
      'PAY_2': 0.0,
      'PAY_3': 0.0,
      'BILL_AMT1': 10000.0,
      'BILL_AMT2': 9500.0,
      'BILL_AMT3': 9000.0,
      'PAY_AMT1': 5000.0,
      'PAY_AMT2': 4750.0,
      'PAY_AMT3': 4500.0,
      '_raw_features': {
        'avg_transaction': 500.0,
        'transaction_volatility': 0.5,
        'expense_ratio': 0.8,
        'payment_consistency': 30.0,
        'overdraft_frequency': 0.0,
        'income_stability': 0.3,
        'category_diversity': 0.5,
        'account_age': 180.0,
        'avg_balance': 5000.0,
      },
    };
  }

  static double _mean(List<double> values) {
    if (values.isEmpty) return 0.0;
    return values.reduce((a, b) => a + b) / values.length;
  }

  static double _standardDeviation(List<double> values) {
    if (values.length < 2) return 0.0;
    final mean = _mean(values);
    final squaredDiffs = values.map((v) => pow(v - mean, 2)).toList();
    return sqrt(squaredDiffs.reduce((a, b) => a + b) / values.length);
  }
}

File: backend/ml/feature_extractor.py
"""Feature extraction from transaction data."""
import pandas as pd
import numpy as np
from typing import Dict, List, Any


class FeatureExtractor:
    """Extract credit-relevant features from transaction data."""

    # Expected columns in transaction CSV
    EXPECTED_COLUMNS = ['date', 'amount', 'category', 'type', 'balance']

    def __init__(self):
        """Initialize the feature extractor."""
        self.features = {}

    def extract_features(self, transactions: List[Dict[str, Any]]) -> Dict[str, float]:
        """
        Extract features from a list of transactions.

        Args:
            transactions: List of transaction dictionaries with keys:
                - date: Transaction date
                - amount: Transaction amount
                - category: Category (groceries, utilities, salary, etc.)
                - type: 'credit' or 'debit'
                - balance: Account balance after transaction

        Returns:
            Dictionary of extracted features mapped to model input format.
        """
        if not transactions:
            return self._get_default_features()

        df = pd.DataFrame(transactions)

        # Convert date to datetime
        df['date'] = pd.to_datetime(df['date'], errors='coerce')
        df['amount'] = pd.to_numeric(df['amount'], errors='coerce').fillna(0)
        df['balance'] = pd.to_numeric(df['balance'], errors='coerce').fillna(0)

        # Sort by date
        df = df.sort_values('date')

        # Extract raw features
        features = {}

        # Income-related features
        income_txns = df[df['type'].str.lower() == 'credit']
        expense_txns = df[df['type'].str.lower() == 'debit']

        total_income = income_txns['amount'].sum() if len(income_txns) > 0 else 0
        total_expenses = expense_txns['amount'].sum() if len(expense_txns) > 0 else 0

        # Helper to handle NaN values
        def safe_float(val, default=0.0):
            if pd.isna(val) or np.isnan(val) if isinstance(val, float) else False:
                return default
            return float(val)

        features['transaction_count'] = len(df)
        features['avg_transaction'] = safe_float(df['amount'].mean(), 500)
        amt_std = df['amount'].std()
        amt_mean = df['amount'].mean()
        features['transaction_volatility'] = safe_float(amt_std / (amt_mean + 1), 0.5) if amt_mean > 0 else 0.5
        features['expense_ratio'] = safe_float(total_expenses / (total_income + 1), 1.0) if total_income > 0 else 1.0

        # Payment consistency (Max days between payments)
        # Standard deviation penalizes freelance/irregular payers.
        # Max gap is better: 35 days = OK, 60 days = Missed a month.
        max_payment_gap = 30.0  # default
        if len(income_txns) > 1:
            income_dates = income_txns['date'].dropna()
            if len(income_dates) > 1:
                # diff() gives days between adjacent payments
                date_diffs = income_dates.diff().dropna().dt.days
                if len(date_diffs) > 0:
                    max_payment_gap = float(date_diffs.max())
        features['payment_consistency'] = max_payment_gap

        # Overdraft frequency (balance going negative)
        features['overdraft_frequency'] = safe_float((df['balance'] < 0).sum() / len(df), 0) if len(df) > 0 else 0

        # Income stability (coefficient of variation of income)
        if len(income_txns) > 1:
            inc_std = income_txns['amount'].std()
            inc_mean = income_txns['amount'].mean()
            features['income_stability'] = safe_float(inc_std / (inc_mean + 1), 0.5)
        else:
            features['income_stability'] = 0.5

        # Category diversity (more diverse = better financial management)
        unique_categories = df['category'].nunique() if 'category' in df.columns else 1
        features['category_diversity'] = min(unique_categories / 10, 1.0)

        # Account age (days between first and last transaction)
        if len(df) > 1 and df['date'].notna().any():
            date_range = (df['date'].max() - df['date'].min()).days
            features['account_age'] = safe_float(date_range, 30)
        else:
            features['account_age'] = 30

        # Average balance
        features['avg_balance'] = safe_float(df['balance'].mean(), 5000)

        # Map to model input format
        return self._map_to_model_features(features, total_income, total_expenses)

    def _map_to_model_features(self, features: Dict[str, float],
                                total_income: float,
                                total_expenses: float) -> Dict[str, float]:
        """
        Map extracted features to the model's expected input format.

        Maps transaction-based features to UCI Credit Card features.
        """
        # Estimate credit limit based on income and stability
        # Stable income (low volatility) suggests higher creditworthiness/limit
        volatility = features.get('transaction_volatility', 0.5)
        
        # Multiplier ranges from 3x (unstable) to 10x (very stable)
        limit_multiplier = max(3.0, 10.0 - (volatility * 10))
        estimated_limit = max(total_income * limit_multiplier, 30000) # Floor at 30k

        # Convert payment consistency (max gap) to payment status
        # PAY_0: -1 = pay on time, 0 = revolving, 1 = late 1 month, etc.
        # -2: No consumption (unused) or full payment
        max_gap = features['payment_consistency']
        
        # Payment Status Mapping
        if max_gap <= 32:
            payment_status = -1  # Paid on time
        elif max_gap <= 40:
            payment_status = 0   # Revolving credit (paid minimum but not full?)
        elif max_gap <= 65:
            payment_status = 1   # 1 month late
        elif max_gap <= 95:
            payment_status = 2   # 2 months late
        else:
            payment_status = min(int(max_gap / 30), 8) # Cap at 8
            
        # Refine payment status with overdraft data
        # If they overdraft frequently, they are likely struggling
        if features.get('overdraft_frequency', 0) > 0.2:
            payment_status = max(payment_status, 2)

        # Calculate bill amounts based on expenses and balance
        # We assume bills are roughly equal to expenses
        bill_amt = total_expenses / 3 if total_expenses > 0 else features['avg_balance'] * 0.3

        # Calculate payment amounts based on income patterns
        # If income > expenses, they likely pay full bill
        # If income < expenses, they pay only a portion
        if total_income > total_expenses:
             pay_amt = bill_amt * 1.05 # Paying full + extra
        else:
             pay_amt = bill_amt * (total_income / (total_expenses + 1))

        model_features = {
            'LIMIT_BAL': estimated_limit,
            'AGE': 35,  # Default age
            'PAY_0': payment_status,
            'PAY_2': payment_status,
            'PAY_3': payment_status,
            'BILL_AMT1': bill_amt,
            'BILL_AMT2': bill_amt * 0.95,
            'BILL_AMT3': bill_amt * 0.9,
            'PAY_AMT1': pay_amt,
            'PAY_AMT2': pay_amt * 0.95,
            'PAY_AMT3': pay_amt * 0.9,
        }

        # Store raw features for explanations
        model_features['_raw_features'] = features

        return model_features

    def _get_default_features(self) -> Dict[str, float]:
        """Return default features when no transactions are provided."""
        return {
            'LIMIT_BAL': 50000,
            'AGE': 35,
            'PAY_0': 0,
            'PAY_2': 0,
            'PAY_3': 0,
            'BILL_AMT1': 10000,
            'BILL_AMT2': 9500,
            'BILL_AMT3': 9000,
            'PAY_AMT1': 5000,
            'PAY_AMT2': 4750,
            'PAY_AMT3': 4500,
            '_raw_features': {
                'avg_transaction': 500,
                'transaction_volatility': 0.5,
                'expense_ratio': 0.8,
                'payment_consistency': 7,
                'overdraft_frequency': 0,
                'income_stability': 0.3,
                'category_diversity': 0.5,
                'account_age': 180,
                'avg_balance': 5000
            }
        }


def parse_csv_transactions(csv_content: str) -> List[Dict[str, Any]]:
    """
    Parse CSV content into a list of transaction dictionaries.

    Args:
        csv_content: CSV string with transaction data

    Returns:
        List of transaction dictionaries
    """
    from io import StringIO

    try:
        df = pd.read_csv(StringIO(csv_content))

        # Normalize column names
        df.columns = df.columns.str.lower().str.strip()

        # Map common column name variations
        column_mapping = {
            'transaction_date': 'date',
            'trans_date': 'date',
            'transaction_amount': 'amount',
            'trans_amount': 'amount',
            'transaction_type': 'type',
            'trans_type': 'type',
            'account_balance': 'balance',
            'running_balance': 'balance',
            'transaction_category': 'category',
            'trans_category': 'category'
        }

        df = df.rename(columns=column_mapping)

        # Ensure required columns exist
        required = ['date', 'amount', 'type']
        for col in required:
            if col not in df.columns:
                # Try to infer from data
                if col == 'type' and 'amount' in df.columns:
                    df['type'] = df['amount'].apply(lambda x: 'credit' if x > 0 else 'debit')
                elif col == 'date':
                    df['date'] = pd.Timestamp.now().strftime('%Y-%m-%d')
                else:
                    df[col] = 0

        # Ensure optional columns exist
        if 'category' not in df.columns:
            df['category'] = 'other'
        if 'balance' not in df.columns:
            df['balance'] = df['amount'].cumsum()

        return df.to_dict('records')

    except Exception as e:
        print(f"Error parsing CSV: {e}")
        return []
